File: bestfitr/src/gev.cpp
// cpp11 glue exposing the GEV slice of the shared C++ core to R.
// The core headers are vendored under src/bestfit_core/include (see tools/sync_core.py).
#include <cpp11.hpp>

#include <string>
#include <vector>

#include "bestfit/numerics/distributions/generalized_extreme_value.hpp"

namespace dist = bestfit::numerics::distributions;
using namespace cpp11;

static dist::EstimationMethod parse_method(const std::string& m) {
    if (m == "mom" || m == "moments") return dist::EstimationMethod::MethodOfMoments;
    if (m == "lmom" || m == "lmoments") return dist::EstimationMethod::MethodOfLinearMoments;
    if (m == "mle") return dist::EstimationMethod::MaximumLikelihood;
    stop("unknown estimation method '%s' (use 'mom', 'lmom', or 'mle')", m.c_str());
}

[[cpp11::register]]
doubles bf_gev_pdf_(doubles x, double location, double scale, double shape) {
    dist::GeneralizedExtremeValue g(location, scale, shape);
    writable::doubles out(x.size());
    for (R_xlen_t i = 0; i < x.size(); ++i) out[i] = g.pdf(x[i]);
    return out;
}

[[cpp11::register]]
doubles bf_gev_cdf_(doubles x, double location, double scale, double shape) {
    dist::GeneralizedExtremeValue g(location, scale, shape);
    writable::doubles out(x.size());
    for (R_xlen_t i = 0; i < x.size(); ++i) out[i] = g.cdf(x[i]);
    return out;
}

[[cpp11::register]]
doubles bf_gev_quantile_(doubles p, double location, double scale, double shape) {
    dist::GeneralizedExtremeValue g(location, scale, shape);
    writable::doubles out(p.size());
    for (R_xlen_t i = 0; i < p.size(); ++i) out[i] = g.inverse_cdf(p[i]);
    return out;
}

[[cpp11::register]]
doubles bf_gev_moments_(double location, double scale, double shape) {
    dist::GeneralizedExtremeValue g(location, scale, shape);
    writable::doubles out({g.mean(), g.median(), g.mode(), g.standard_deviation(), g.skewness(),
                           g.kurtosis(), g.minimum(), g.maximum()});
    out.names() = {"mean", "median", "mode", "sd", "skewness", "kurtosis", "minimum", "maximum"};
    return out;
}

[[cpp11::register]]
bool bf_gev_valid_(double location, double scale, double shape) {
    return dist::GeneralizedExtremeValue(location, scale, shape).parameters_valid();
}

[[cpp11::register]]
doubles bf_gev_fit_(doubles data, std::string method) {
    std::vector<double> sample(data.begin(), data.end());
    dist::GeneralizedExtremeValue g;
    g.estimate(sample, parse_method(method));
    writable::doubles out({g.xi(), g.alpha(), g.kappa()});
    out.names() = {"location", "scale", "shape"};
    return out;
}

[[cpp11::register]]
double bf_gev_quantile_variance_(double p, double location, double scale, double shape,
                                 int sample_size) {
    dist::GeneralizedExtremeValue g(location, scale, shape);
    return g.quantile_variance(p, sample_size);
}

[[cpp11::register]]
doubles bf_gev_linear_moments_(double location, double scale, double shape) {
    dist::GeneralizedExtremeValue g(location, scale, shape);
    auto lmom = g.linear_moments_from_parameters({location, scale, shape});
    return writable::doubles(lmom);
}

[[cpp11::register]]
doubles bf_gev_quantile_gradient_(double p, double location, double scale, double shape) {
    dist::GeneralizedExtremeValue g(location, scale, shape);
    return writable::doubles(g.quantile_gradient(p));
}

// Returns the flattened 3x3 covariance (row-major).
[[cpp11::register]]
doubles bf_gev_parameter_covariance_(double location, double scale, double shape,
                                     int sample_size) {
    dist::GeneralizedExtremeValue g(location, scale, shape);
    auto cov = g.parameter_covariance(sample_size);
    writable::doubles out(9);
    for (int i = 0; i < 3; ++i)
        for (int j = 0; j < 3; ++j) out[i * 3 + j] = cov[i][j];
    return out;
}

File: bestfitr/R/gev.R
#' Generalized Extreme Value distribution
#'
#' Density, distribution, and quantile functions for the Generalized Extreme
#' Value (GEV) distribution, parameterised by location \eqn{\xi}, scale
#' \eqn{\alpha}, and shape \eqn{\kappa}. These wrap the shared C++ core ported
#' from the USACE-RMC Numerics library.
#'
#' @param x,q numeric vector of quantiles.
#' @param p numeric vector of probabilities.
#' @param location location parameter \eqn{\xi}.
#' @param scale scale parameter \eqn{\alpha} (must be positive).
#' @param shape shape parameter \eqn{\kappa}.
#'
#' @return A numeric vector the same length as `x`, `q`, or `p`.
#' @name gev
#' @examples
#' dgev(100, location = 100, scale = 10, shape = 0)
#' qgev(0.99, location = 10849, scale = 5745.6, shape = 0.005)
NULL

#' @rdname gev
#' @export
dgev <- function(x, location = 0, scale = 1, shape = 0) {
  bf_gev_pdf_(as.double(x), location, scale, shape)
}

#' @rdname gev
#' @export
pgev <- function(q, location = 0, scale = 1, shape = 0) {
  bf_gev_cdf_(as.double(q), location, scale, shape)
}

#' @rdname gev
#' @export
qgev <- function(p, location = 0, scale = 1, shape = 0) {
  bf_gev_quantile_(as.double(p), location, scale, shape)
}

#' Moments of a GEV distribution
#'
#' @inheritParams gev
#' @return A named numeric vector: `mean`, `median`, `sd`, `skewness`,
#'   `kurtosis`, `minimum`, `maximum`. Entries are `NaN` where the moment is
#'   undefined for the given shape.
#' @export
#' @examples
#' gev_moments(location = 100, scale = 10, shape = 0)
gev_moments <- function(location, scale, shape) {
  bf_gev_moments_(location, scale, shape)
}

#' Fit a GEV distribution
#'
#' Estimate GEV parameters from a sample.
#'
#' @param x numeric vector of observations.
#' @param method estimation method: `"mle"` (maximum likelihood, the default),
#'   `"lmom"` (L-moments), or `"mom"` (product moments).
#' @return A named numeric vector with `location`, `scale`, and `shape`.
#' @export
#' @examples
#' set.seed(1)
#' gev_fit(rnorm(100, 100, 10), method = "lmom")
gev_fit <- function(x, method = c("mle", "lmom", "mom")) {
  method <- match.arg(method)
  bf_gev_fit_(as.double(x), method)
}

#' Standard error of a GEV quantile (maximum likelihood)
#'
#' Delta-method standard error of the `p`-quantile, using the expected
#' (Fisher) information matrix.
#'
#' @inheritParams gev
#' @param sample_size the sample size the fit was based on.
#' @return The standard error (a single numeric value).
#' @export
#' @examples
#' gev_quantile_se(0.99, location = 10849, scale = 5745.6, shape = 0.005,
#'                 sample_size = 62)
gev_quantile_se <- function(p, location, scale, shape, sample_size) {
  sqrt(bf_gev_quantile_variance_(p, location, scale, shape, as.integer(sample_size)))
}
